File: leetcode/1710-maximum-units-on-a-truck/solution.cpp
int maximumUnits(vector<vector<int>>& boxTypes, int truckSize) {
    vector<pair<int, int>> boxes;

    for (int i = 0; i < boxTypes.size(); i++) {
        boxes.push_back(make_pair(boxTypes[i][1], boxTypes[i][0]));
    }

    sort(boxes.begin(), boxes.end(), greater<pair<int, int>>());

    int units = 0;
    int i = 0;
    while (truckSize > 0 && i < boxes.size()) {
        if (boxes[i].second >= truckSize) {
            units += truckSize * boxes[i].first;
            truckSize = 0;
            break;
        }

        if (boxes[i].second < truckSize) {
            units += boxes[i].second * boxes[i].first;
            truckSize -= boxes[i].second;
        }

        i++;
    }

    return units;
}

File: leetcode/1710-maximum-units-on-a-truck/solution.ts
function maximumUnits(boxTypes: number[][], truckSize: number): number {
  const boxes: number[][] = [];

  for (let i = 0; i < boxTypes.length; i++) {
    boxes.push([boxTypes[i][1], boxTypes[i][0]]);
  }

  boxes.sort(function (a: number[], b: number[]): number {
    return b[0] - a[0];
  });

  let units = 0;
  let i = 0;

  while (truckSize > 0 && i < boxes.length) {
    if (boxes[i][1] >= truckSize) {
      units += truckSize * boxes[i][0];
      truckSize = 0;
      break;
    }

    units += boxes[i][1] * boxes[i][0];
    truckSize -= boxes[i][1];

    i++;
  }

  return units;
}
